File: ui/src/types/member.ts
import type { IntegrationMember } from "./integration";
import type { User } from "./user";

export enum Role {
  Maintainer = "maintainer",
  Owner = "owner",
  Reader = "reader",
  Writer = "writer",
}

export type UserMember = {
  userId: string;
  role: Role;
  user?: User;
};

export type Member = UserMember | IntegrationMember;

File: server/api/pkg/workspace/member.go
package workspace

type Member interface {
	GetRole() Role
	GetMemberType() string
}

type User struct {
	Name  string
	Email string
	ID    UserID
}

type UserMember struct {
	Host   *string
	User   *User
	Role   Role
	UserID UserID
}

type IntegrationMember struct {
	InvitedBy     *User
	Role          Role
	IntegrationID IntegrationID
	InvitedByID   UserID
	Active        bool
}

func (m UserMember) GetRole() Role         { return m.Role }
func (m UserMember) GetMemberType() string { return "UserMember" }

func (m IntegrationMember) GetRole() Role         { return m.Role }
func (m IntegrationMember) GetMemberType() string { return "IntegrationMember" }
